kernel/src/protocols/attestation.rs

```diff
@@ -0,0 +1,356 @@
+        let size = self.nonce_size as usize;
+
+        Ok((gpa, size))
+    }
+
+    /// Returns the manifest GPA and size
+    /// Checks if gpa is page aligned, valid  and does not cross page boundary.
+    pub fn get_manifest_gpa_and_size(&self) -> Result<(PhysAddr, usize), SvsmReqError> {
+        let gpa = PhysAddr::from(self.manifest_gpa);
+        if !gpa.is_aligned(8) || !valid_phys_address(gpa) || gpa.crosses_page(8) {
+            return Err(SvsmReqError::invalid_parameter());
+        }
+
+        // Won't panic on amd64 as usize > u32 always
+        let size = self.manifest_size as usize;
+
+        Ok((gpa, size))
+    }
+
+    /// Returns the guid
+    /// Checks if gpa is page aligned, valid  and does not cross page boundary
+    pub fn get_report_gpa_and_size(&self) -> Result<(PhysAddr, usize), SvsmReqError> {
+        let gpa = PhysAddr::from(self.report_gpa);
+        if !gpa.is_aligned(8) || !valid_phys_address(gpa) || gpa.crosses_page(8) {
+            return Err(SvsmReqError::invalid_parameter());
+        }
+
+        // Won't panic on amd64 as usize > u32 always
+        let size = self.report_size as usize;
+
+        Ok((gpa, size))
+    }
+
+    /// Returns the certificate GPA and size
+    /// Checks if gpa is page aligned, valid  and does not cross page boundary
+    /// pub fn get_report_gpa_and_size(&self) -> Result<(PhysAddr, usize), SvsmReqError> {
+    /// Currently not implemented supported as no service attestation currently supports
+    /// returning certificates. Implement when needed by copying get_report_gpa_and_size()
+    /// and replacing report's gpa and size  with certificate's.
+    /// }
+
+    pub fn get_manifest_version(&self) -> u32 {
+        self.manifest_ver
+    }
+
+    fn is_manifest_version_valid(&self) -> bool {
+        //Currently only manifest version 0 is supported
+        self.manifest_ver == 0
+    }
+    pub fn get_guid(&self) -> u128 {
+        u128::from_le_bytes(self.guid)
+    }
+}
+
+#[cfg(all(feature = "vtpm", not(test)))]
+fn get_attestation_report(nonce: &[u8]) -> Result<Vec<u8>, SvsmReqError> {
+    //Construct attestation request message to send to SNP
+    let mut report_req = Vec::<u8>::with_capacity(size_of::<SnpReportResponse>());
+    let mut buf = Vec::<u8>::with_capacity(USER_DATA_SIZE);
+    buf.fill(0);
+
+    if nonce.len() > USER_DATA_SIZE {
+        // If the nonce is greater than the user data size, return an error as something is wrong.
+        return Err(SvsmReqError::invalid_parameter());
+    }
+    // Copy user attestation request nonce to buffer
+    buf.extend_from_slice(nonce);
+    report_req.extend_from_slice(&buf[..nonce.len()]);
+
+    // Set request VMPL to 0
+    report_req.extend_from_slice(&0_u32.to_le_bytes());
+
+    // Set reserved bytes to zeros
+    report_req.extend_from_slice(&[0; 28]);
+
+    // Make sure buffer is big enough to hold the report
+    report_req.resize(size_of::<SnpReportResponse>(), 0);
```

Comment:
I'd do for sure in a separate commit, but this PR looks the right place since it's the user of that API. BTW not a strong opinion, if you prefer another PR is fine by me.